Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy.swift

```diff
@@ -102,9 +102,13 @@ struct PathHierarchy {
             var nodes: [String: Node] = [:]
             nodes.reserveCapacity(graph.symbols.count)
             for (id, symbol) in graph.symbols {
-                let node = Node(symbol: symbol)
-                nodes[id] = node
-                allNodes[id, default: []].append(node)
+                if let existingNode = allNodes[id]?.first(where: { $0.symbol!.identifier == symbol.identifier }) {
```

Comment:
Are we sure the force unwrap is okay here? I'm guessing yes since we're iterating over `graph.symbols` here but I wonder if then we should have a specific kind of `SymbolNode` where this is non-optional? (not for this PR just wondering generally)

CC: @d-ronnqvist

Reply:
Separately – do we expect the list of existing `allNodes[id]` to always be a trivially short list or is there potential for a noticeable performance penalty for iterating here?

Reply:
> Are we sure the force unwrap is okay here? I'm guessing yes since we're iterating over `graph.symbols` here but I wonder if then we should have a specific kind of `SymbolNode` where this is non-optional? (not for this PR just wondering generally)

Yes, because all the non-symbols are added after the initial tree has been built it's safe to force unwrap the symbol in the initializer code that builds up the initial tree. 

Maybe we should make the distinction between symbol nodes and non-symbol nodes clearer in the `PathHierarchy` implementation. That should be the only place that needs to know the difference. 

There's only a handful of places in `PathHierarchyBasedLinkResolver` that access nodes directly. I think that by adding a few targeted helper functions to accomplish the same things we could probably make the `Node` completely private to the `PathHierarchy` where the the behaviors are well defined and (mostly) covered by debug assertions. 

Either of these changes would definitely be for a different PR.

> Separately – do we expect the list of existing allNodes[id] to always be trivially short or is there potential for a noticeable performance penalty for iterating here?

Yes, the symbol USRs are unique per symbol graph file so in the worst case `allNodes[id]` will be as long as the number of main symbol graph files for that module. In practice that means that the worst case is _n<sub>languages</sub>_ × _n<sub>platforms</sub>_ which is still a single digit at worst today.